src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,662 @@
+		{
+			try
+			{
+				if( now )
+				{
+					this.gridProxy.getTick().alertDevice( this.gridProxy.getNode() );
+				}
+				else
+				{
+					this.gridProxy.getTick().sleepDevice( this.gridProxy.getNode() );
+				}
+			}
+			catch( final GridAccessException e )
+			{
+				// :P
+			}
+		}
+	}
+
+	@Override
+	public void onChangeInventory( IItemHandler inv, int slot, InvOperation mc, ItemStack removedStack, ItemStack newStack )
+	{
+		if( this.isWorking == slot )
+		{
+			return;
+		}
+
+		if( inv == this.config )
+		{
+			this.readConfig();
+		}
+	}
+
+	@Override
+	public int getPriority()
+	{
+		return this.priority;
+	}
+
+	@Override
+	public void setPriority( final int newValue )
+	{
+		this.priority = newValue;
+	}
+
+	public void writeToNBT( final NBTTagCompound data )
+	{
+		final NBTTagList tankContents = new NBTTagList();
+		for( int i = 0; i < NUMBER_OF_TANKS; ++i )
+		{
+			tankContents.appendTag( this.tanks[i].writeToNBT( new NBTTagCompound() ) );
+		}
+
+		data.setInteger( "priority", this.priority );
+		data.setTag( "storage", tankContents );
+		this.config.writeToNBT( data, "config" );
+	}
+
+	public void readFromNBT( final NBTTagCompound data )
+	{
+		this.config.readFromNBT( data, "config" );
+		final NBTTagList tankContents = data.getTagList( "storage", 10 );
+		if( tankContents != null )
+		{
+			for( int i = 0; i < Math.min( NUMBER_OF_TANKS, tankContents.tagCount() ); ++i )
+			{
+				this.tanks[i].readFromNBT( tankContents.getCompoundTagAt( i ) );
+			}
+		}
+		this.priority = data.getInteger( "priority" );
+		this.readConfig();
+	}
+
+	public IItemHandler getConfig()
+	{
+		return this.config;
+	}
+
+	public AEFluidTank getTank( final int i )
+	{
+		return this.tanks[i];
+	}
+
+	public boolean writeTankInfo( final Map<Integer, NBTTagCompound> tagMap, boolean all )
+	{
+		boolean empty = true;
+		for( int i = 0; i < NUMBER_OF_TANKS; ++i )
+		{
+			if( all || this.tankChanged[i] )
+			{
+				tagMap.put( i, this.tanks[i].writeToNBT( new NBTTagCompound() ) );
+				this.tankChanged[i] = !all;
+				empty = false;
+			}
+		}
+		return !empty;
+	}
+
+	public boolean readTankInfo( final Map<Integer, NBTTagCompound> tagMap )
+	{
+		boolean changed = false;
+		for( int i = 0; i < NUMBER_OF_TANKS; ++i )
+		{
+			if( tagMap.containsKey( i ) )
+			{
+				this.tanks[i].readFromNBT( tagMap.get( i ) );
+				changed = true;
+			}
+		}
+		return changed;
+	}
+
+	private class InterfaceRequestSource extends MachineSource
+	{
+		private final InterfaceRequestContext context;
+
+		InterfaceRequestSource( IActionHost v )
+		{
+			super( v );
+			this.context = new InterfaceRequestContext();
+		}
+
+		@Override
+		public <T> Optional<T> context( Class<T> key )
+		{
+			if( key == InterfaceRequestContext.class )
+			{
+				return (Optional<T>) Optional.of( this.context );
+			}
+
+			return super.context( key );
+		}
+	}
+
+	private class InterfaceRequestContext implements Comparable<Integer>
+	{
+		@Override
+		public int compareTo( Integer o )
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Override "equals(Object obj)" to comply with the contract of the "compareTo(T o)" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1210)